agent/dehydrate.js

```diff
@@ -30,63 +84,72 @@
  * }
  * and cleaned = [["some", "attr"], ["other"]]
  */
-function dehydrate(data: Object, cleaned: Array<Array<string>>, path?: Array<string>, level?: number): string | Object {
-  level = level || 0;
-  path = path || [];
-  if (typeof data === 'function') {
-    cleaned.push(path);
-    return {
-      name: data.name,
-      type: 'function',
-    };
-  }
-  if (!data || typeof data !== 'object') {
-    if (typeof data === 'string' && data.length > 500) {
-      return data.slice(0, 500) + '...';
-    }
+function dehydrate(data: Object, cleaned: Array<Array<string>>, path?: Array<string> = [], level?: number = 0): string | Object {
+
+  var type = getPropType(data);
+
+  switch (type) {
+
+    case 'function':
+      cleaned.push(path);
+      return {
+        name: data.name,
+        type: 'function',
+      };
+
+    case 'string':
+      return data.length <= 500 ? data : data.slice(0, 500) + '...';
+
     // We have to do this assignment b/c Flow doesn't think "symbol" is
     // something typeof would return. Error 'unexpected predicate "symbol"'
-    var type = typeof data;
-    if (type === 'symbol') {
+    case 'symbol':
       cleaned.push(path);
       return {
         type: 'symbol',
         name: data.toString(),
       };
-    }
-    return data;
-  }
-  if (data._reactFragment) {
+
     // React Fragments error if you try to inspect them.
-    return 'A react fragment';
-  }
-  if (level > 2) {
-    cleaned.push(path);
-    return {
-      type: Array.isArray(data) ? 'array' : 'object',
-      name: !data.constructor || data.constructor.name === 'Object' ? '' : data.constructor.name,
-      meta: Array.isArray(data) ? {
-        length: data.length,
-      } : null,
-    };
-  }
-  if (Array.isArray(data)) {
-    // $FlowFixMe path is not undefined.
-    return data.map((item, i) => dehydrate(item, cleaned, path.concat([i]), level + 1));
-  }
-  // TODO when this is in the iframe window, we can just use Object
-  if (data.constructor && typeof data.constructor === 'function' && data.constructor.name !== 'Object') {
-    cleaned.push(path);
-    return {
-      name: data.constructor.name,
-      type: 'object',
-    };
-  }
-  var res = {};
-  for (var name in data) {
-    res[name] = dehydrate(data[name], cleaned, path.concat([name]), level + 1);
+    case 'react_fragment':
+      return 'A React Fragment';
+
+    // ArrayBuffers error if you try to inspect them.
+    case 'array_buffer':
+    case 'data_view':
+      cleaned.push(path);
+      return {
+        type,
+        name: type === 'data_view' ? 'DataView' : 'ArrayBuffer',
+        meta: {
+          length: data.byteLength,
+          uninspectable: true,
+        },
+      };
+
+    case 'array':
+      if (level > 2) {
+        return createDehydrated(type, data, cleaned, path);
+      }
+      return data.map((item, i) => dehydrate(item, cleaned, path.concat([i]), level + 1));
```

Comment:
this would also be nice to limit to e.g. 100, but I'm fine w/ just putting a TODO